### PR TITLE
Code Review for GNU Radio Block Header Parsing tool.

### PR DESCRIPTION
**Pull request by aru31**
Code Review for GNU Radio Block Header Parsing tool

### CODE REVIEW

File: gr-utils/python/blocktool/core/iosignature.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+        "output": {
+            "signature": None
+        }
+    }
+    with open(io_file, 'r') as impl:
+        lines = []
+        for line in impl:
+            if Constants.IO_SIGNATURE in line:
+                lines.append(line)
+    impl.close()
```
Comment (primercuervo) [inline gr-utils/python/blocktool/core/iosignature.py]: When using `with open`, the file gets automatically closed after the `with` block thanks to the context manager, so line 54 is not required.
```suggestion
```

File: gr-utils/python/blocktool/core/parseheader.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        self.info['impl_dir'] = os.path.abspath(os.path.join(file_path,
+                                                             '..', '..', '..', '..', 'lib'))
+        self.info['impl_file'] = os.path.join(self.info['impl_dir'],
+                                              self.info['filename'].split('.')[0]+'_impl.cc')
+        self.validate()
+
+    def validate(self):
+        """ Override the Blocktool validate function """
+        BlockTool._validate(self)
+        self._validate()
```
Comment (primercuervo) [inline gr-utils/python/blocktool/core/parseheader.py]: I'm a bit confused about L90-L91. Aren't both lines up to this point supposed to be just `pass`?

File: gr-utils/python/blocktool/core/parseheader.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+                        }
+                        for argument in setter.arguments:
+                            args = {
+                                "name": str(argument.name),
+                                "dtype": str(argument.decl_type)
+                            }
+                            getter_arguments.append(args['name'])
+                            setter_args['arguments_type'].append(args.copy())
+                        self.parsed_data['methods'].append(setter_args.copy())
+        except:
```
Comment (primercuervo) [inline gr-utils/python/blocktool/core/parseheader.py]: Wouldn't it be a bit obscure for the API user if the exception doesn't `raises` or produce a type of input to work with? it seems that when the getters and setters fail we end up, unknowingly, with empty parsed data

File: gr-utils/python/blocktool/tests/test_blocktool.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        self.target_file = os.path.abspath(
+            os.path.join(self.test_dir, 'agc2_cc.h'))
+        self.impl_dir = os.path.abspath(os.path.join(self.current_folder,
+                                                     '../../../../gr-analog/lib'))
+        self.modname = os.path.basename(os.path.dirname(
+            os.path.dirname(os.path.dirname(os.path.dirname(self.target_file)))))
+        self.gr = self.modname.split('-')[0]
+        self.module = self.modname.split('-')[-1]
+
+    def is_int(self, number):
```
Comment (primercuervo) [inline gr-utils/python/blocktool/tests/test_blocktool.py:67]: this is a general UnitTest but seems unrelated to the specifics of BlockTool

File: gr-utils/python/blocktool/core/parseheader.py (hunk 2 above)
Comment (aru31) [reply gr-utils/python/blocktool/core/parseheader.py]: L90 and 91 points to the same validate function in `core/base.py`, sorry for the repetition. They are a pass, for now, but if blocktool gets additional functionality in future, common blocktool validation in `core/base.py` might come in handy.

File: gr-utils/python/blocktool/core/parseheader.py (hunk 3 above)
Comment (aru31) [reply gr-utils/python/blocktool/core/parseheader.py]: This was one of the doubts I had and wanted to clear with you. Sure, an exception can be raised.

File: gr-utils/python/blocktool/tests/test_blocktool.py (hunk 4 above)
Comment (aru31) [reply gr-utils/python/blocktool/tests/test_blocktool.py:67]: I didn't get it. All the __init__ variables defined above are required to check the Blocktool raised exceptions as in the validate function of `core/parseheader.py`

File: gr-utils/python/blocktool/__init__.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with GNU Radio; see the file COPYING.  If not, write to
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+
+from .core.parseheader import BlockHeaderParser
+from .core.outputschema import PARSED_OUTPUT_SCHEME
```
Comment (skoslowski) [inline gr-utils/python/blocktool/__init__.py]: I prefer s.th. like RESULT_SCHEMA

File: gr-utils/python/blocktool/core/parseheader.py [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+from __future__ import print_function
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+import os
+import re
+import codecs
+import logging
+
+import click
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/parseheader.py]: No cli stuff in the core, please

File: gr-utils/python/blocktool/core/parseheader.py [hunk 7]
```diff
@@ -0,0 +1,281 @@
...
+
+from pygccxml import parser
+from pygccxml import declarations
+from pygccxml import utils
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/parseheader.py]: collapse?

File: gr-utils/python/blocktool/core/parseheader.py [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+        file_path = os.path.abspath(file_path)
+        self.info['target_file'] = file_path
+        self.info['modname'] = os.path.basename(os.path.dirname(
+            os.path.dirname(os.path.dirname(os.path.dirname(file_path)))))
+        self.info['filename'] = os.path.basename(file_path)
+        self.info['target_dir'] = os.path.dirname(file_path)
+        self.info['impl_dir'] = os.path.abspath(os.path.join(file_path,
+                                                             '..', '..', '..', '..', 'lib'))
+        self.info['impl_file'] = os.path.join(self.info['impl_dir'],
+                                              self.info['filename'].split('.')[0]+'_impl.cc')
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/parseheader.py]: why aren't these defined as attributes of the instance? want benefit does "info" have

File: gr-utils/python/blocktool/core/parseheader.py [hunk 9]
```diff
@@ -0,0 +1,281 @@
...
+                                   self.info['modname'],
+                                   'include',
+                                   'gnuradio',
+                                   self.info['modname'].split('-')[-1])
+        if _target_dir != self.info['target_dir']:
+            raise BlockToolException('Invalid GNU Radio module')
+        try:
+            open(self.info['impl_file'], 'r')
+        except OSError:
+            raise OSError
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/parseheader.py]: What is this for?

File: gr-utils/python/blocktool/core/parseheader.py [hunk 10]
```diff
@@ -0,0 +1,281 @@
...
+        if not self.info['filename'].endswith('.h'):
+            raise BlockToolException(
+                'Invalid header file')
+        _target_dir = os.path.join(self.target_dir,
+                                   self.info['modname'],
+                                   'include',
+                                   'gnuradio',
+                                   self.info['modname'].split('-')[-1])
+        if _target_dir != self.info['target_dir']:
+            raise BlockToolException('Invalid GNU Radio module')
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/parseheader.py]: Explain why in the message

File: gr-utils/python/blocktool/core/parseheader.py [hunk 11]
```diff
@@ -0,0 +1,281 @@
...
+            self.parsed_data['namespace'] = [gr, module]
+            if main_namespace.declarations:
+                for _namespace in main_namespace.declarations:
+                    if isinstance(_namespace, declarations.namespace_t):
+                        if Constants.KERNEL not in str(_namespace):
+                            main_namespace = _namespace
+                            self.parsed_data['namespace'].append(
+                                str(_namespace).split('::')[-1].split(' ')[0])
+        except BlockToolException as exception:
+            raise BlockToolException(
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/parseheader.py:130]: This looks weird to me, shouldn't this be another type of exception

File: gr-utils/python/blocktool/tests/test_blocktool.py [hunk 12]
```diff
@@ -0,0 +1,187 @@
...
+            BlockHeaderParser(**test_dict).run
+        # test for valid header file
+        test_dict['file_path'] = os.path.abspath(
+            os.path.join(self.test_dir, 'CMakeLists.txt'))
+        if not os.path.basename(test_dict['file_path']).endswith('.h'):
+            with self.assertRaises(BlockToolException):
+                BlockHeaderParser(**test_dict).run
+        # test for invalid header module
+        test_dict['file_path'] = os.path.abspath(
+            os.path.join(self.current_folder, '../../../../gr-utils/python/blocktool/__init__.py'))
```
Comment (skoslowski) [inline gr-utils/python/blocktool/tests/test_blocktool.py]: ???

File: gr-utils/python/blocktool/tests/test_blocktool.py [hunk 13]
```diff
@@ -0,0 +1,187 @@
...
+            warnings.simplefilter("ignore", ResourceWarning)
+        except NameError:
+            pass
+
+        # test for all the Blocktool Exceptions
+        test_dict = {}
+        # test for none type file_path
+        test_dict['file_path'] = None
+        with self.assertRaises(TypeError):
+            BlockHeaderParser(**test_dict).run
```
Comment (skoslowski) [inline gr-utils/python/blocktool/tests/test_blocktool.py]: missing ()?

File: gr-utils/python/gr_blocktool [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+#
+# You should have received a copy of the GNU General Public License
+# along with GNU Radio; see the file COPYING.  If not, write to
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+#
+""" A tool for parsing header files. """
+
+
+from blocktool.cli.base import cli
```
Comment (skoslowski) [inline gr-utils/python/gr_blocktool]: pull the cli into the blocktool package

shouldn't this be a subpackage of gnuradio

File: gr-utils/python/blocktool/core/parseheader.py [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+    module_types = tuple(module_types)
+    parsed_data = {}
+
+    def __init__(self, file_path=None, **kwargs):
+        """ __init__ """
+        BlockTool.__init__(self, **kwargs)
+        try:
+            open(file_path, 'r')
+        except TypeError:
+            raise TypeError
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/parseheader.py]: ?

**Review by skoslowski — COMMENTED** [review]
high-level feedback (let's consider this part1 of my review)

File: gr-utils/python/blocktool/tests/test_blocktool.py [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+
+
+class TestBlocktoolCore(unittest.TestCase):
+    """ The Tests for blocktool core """
+
+    def __init__(self, *args, **kwargs):
+        super(TestBlocktoolCore, self).__init__(*args, **kwargs)
+        self.current_folder = os.path.abspath(os.path.dirname(__file__))
+        self.test_dir = os.path.abspath(os.path.join(
+            self.current_folder, '../../../../gr-analog/include/gnuradio/analog'))
```
Comment (skoslowski) [inline gr-utils/python/blocktool/tests/test_blocktool.py]: may define a top-level variable that contains the path to the GR source root

File: gr-utils/python/gr_blocktool (hunk 14 above)
Comment (aru31) [reply gr-utils/python/gr_blocktool]: It definitely is, this is just a temporary script for testing purposes only.

File: gr-utils/python/blocktool/core/parseheader.py (hunk 10 above)
Comment (aru31) [reply gr-utils/python/blocktool/core/parseheader.py]: I have now cleaned the exceptions.

File: gr-utils/python/blocktool/tests/test_blocktool.py (hunk 12 above)
Comment (aru31) [reply gr-utils/python/blocktool/tests/test_blocktool.py]: This was a debug statement. Sorry for the mess. I have now cleaned up the tests.

File: gr-utils/python/blocktool/core/parseheader.py (hunk 11 above)
Comment (aru31) [reply gr-utils/python/blocktool/core/parseheader.py:130]: Changed to RuntimeError

File: gr-utils/python/blocktool/cli/CMakeLists.txt [hunk 17]
```diff
@@ -0,0 +1,27 @@
...
+# the Free Software Foundation, Inc., 51 Franklin Street,
+# Boston, MA 02110-1301, USA.
+
+include(GrPython)
+
+GR_PYTHON_INSTALL(FILES
+    __init__.py
+    base.py
+    parseheader.py
+    DESTINATION ${GR_PYTHON_DIR}/gnuradio/blocktool/cli
```
Comment (primercuervo) [inline gr-utils/python/blocktool/cli/CMakeLists.txt]: @skoslowski do we want to install the cli files? Or should we just leave them to be run locally from the source code? I assume the latter..

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 18]
```diff
@@ -0,0 +1,299 @@
...
+from ..core import Constants
+
+LOGGER = logging.getLogger(__name__)
+
+## setup dumper for dumping OrderedDict ##
+_MAPPING_TAG = yaml.resolver.BaseResolver.DEFAULT_MAPPING_TAG
+
+
+def dict_representer(dumper, data):
+    """ Representer to represent special OrderedDict """
```
Comment (primercuervo) [inline gr-utils/python/blocktool/cli/parseheader.py]: "Representer to represent..." seems a bit redundant

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 19]
```diff
@@ -0,0 +1,299 @@
...
+        for _output in self.parsed_data['message_port']['output']:
+            m_output_sig = OrderedDict()
+            m_output_sig['domain'] = 'message'
+            m_output_sig['id'] = _output
+            output_signature.append(m_output_sig)
+    if output_signature:
+        data['outputs'] = output_signature
+
+    _cpp_templates = [('includes', '#include <gnuradio/{}/{}>'.format(block, self.filename)),
+                      ('declartions', '{}::{}::sptr ${{id}}'.format(block, header)),
```
Comment (primercuervo) [inline gr-utils/python/blocktool/cli/parseheader.py]: typo in "declarations"

File: gr-utils/python/blocktool/core/parseheader.py [hunk 20]
```diff
@@ -0,0 +1,278 @@
...
+            self.parsed_data['namespace'] = [gr, module]
+            if main_namespace.declarations:
+                for _namespace in main_namespace.declarations:
+                    if isinstance(_namespace, declarations.namespace_t):
+                        if Constants.KERNEL not in str(_namespace):
+                            main_namespace = _namespace
+                            self.parsed_data['namespace'].append(
+                                str(_namespace).split('::')[-1].split(' ')[0])
+        except RuntimeError:
+            raise BlockToolException(
```
Comment (primercuervo) [inline gr-utils/python/blocktool/core/parseheader.py:130]: Why re-raising and not just handling the `RuntimeError` as is? As far as I understand, the `BlockToolException` is not specialized but just a wrapper for the general `exception`

**Review by primercuervo — CHANGES_REQUESTED** [review]
Just a few comments on the fly that I got to do on a free timeslot. I will add more comments tonight

File: gr-utils/python/blocktool/cli/parseheader.py (hunk 18 above)
Comment (aru31) [reply gr-utils/python/blocktool/cli/parseheader.py]: Representing with Ordered dict and then using it to convert in YAML parameters is quite easy. The whole blocktool yaml_generator is based on this.

File: gr-utils/python/blocktool/core/outputschema.py [hunk 21]
```diff
@@ -0,0 +1,169 @@
...
+""" Schema to be strictly followed be parsed header output """
+
+
+RESULT_SCHEMA = {
+    "title": "JSON SCHEMA TO BE FOLLOWED BY BLOCK HEADER PARSING TOOL",
+    "description": "Schema designed for the header file parsed python dict output",
+    "type": "object",
+    "properties": {
+        "namespace": {
+            "description": "List of nested namspaces",
```
Comment (primercuervo) [inline gr-utils/python/blocktool/core/outputschema.py]: typo "namespaces"

File: gr-utils/python/blocktool/core/parseheader.py [hunk 22]
```diff
@@ -0,0 +1,281 @@
...
+                            args = {
+                                "name": str(argument.name),
+                                "dtype": str(argument.decl_type)
+                            }
+                            getter_arguments.append(args['name'])
+                            setter_args['arguments_type'].append(args.copy())
+                        self.parsed_data['methods'].append(setter_args.copy())
+            else:
+                self.parsed_data['methods'] = []
+        except RuntimeError:
```
Comment (primercuervo) [inline gr-utils/python/blocktool/core/parseheader.py:232]: so the result from not having `setters` (or `getters` for lines 238 on) is the same as when there is a `RuntimeError`. Is this on purpose? Any specific reason not to re-raise with your custom exception and/or notify that an exception happened?

File: gr-utils/python/blocktool/tests/test_blocktool.py [hunk 23]
```diff
@@ -0,0 +1,199 @@
...
+                if not port['id']:
+                    valid_input_message_port = False
+        if output_port:
+            for port in output_port:
+                if not port['id']:
+                    valid_output_message_port = False
+        self.assertTrue(valid_input_message_port)
+        self.assertTrue(valid_output_message_port)
+
+    @unittest.skipIf(SKIP_BLOCK_TEST, 'pygccxml not founf, skipping this unittest')
```
Comment (primercuervo) [inline gr-utils/python/blocktool/tests/test_blocktool.py]: Typo "found"

File: gr-utils/python/blocktool/core/parseheader.py (hunk 22 above)
Comment (aru31) [reply gr-utils/python/blocktool/core/parseheader.py:232]: Yeah. The else statement is not required. I have made the changes.

File: gr-utils/python/blocktool/cli/parseheader.py (hunk 18 above)
Comment (primercuervo) [reply gr-utils/python/blocktool/cli/parseheader.py]: I'm not referring to the implementation or return, but to the grammar of the docstring :)

File: gr-utils/python/blocktool/cli/CMakeLists.txt (hunk 17 above)
Comment (skoslowski) [reply gr-utils/python/blocktool/cli/CMakeLists.txt]: I say, install. At least to have a "python -m gnuradio.blocktool". 

therefore blocktool needs a __main__ module
https://docs.python.org/3/using/cmdline.html#cmdoption-m

I'd start with
```
import sys
from .cli.base import cli

sys.exit(cli())
```
we could then also consider removing gr_blocktool

File: gr-utils/python/blocktool/cli/base.py [hunk 24]
```diff
@@ -0,0 +1,126 @@
...
+
+@with_plugins(iter_entry_points('gnuradio.blocktool.cli.plugins'))
+@click.command(
+    cls=CommandCLI,
+    epilog='A tool to parse block header files. ' +
+    'Call it without options to run specified command interactively'
+)
+def cli():
+    """Block header parsing tool."""
+    pass
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/base.py]: IMO, plugin support and subcommands are too much. Especially if there is only one...
I propose to strip all the boilerplate stuff and trim this down to one module `cli.py` 

click is fine, but I don't see a need for plugins...

File: gr-utils/python/blocktool/cli/base.py [hunk 25]
```diff
@@ -0,0 +1,126 @@
...
+
+    def get_command(self, ctx, cmd_name):
+        """
+        Returns a command object if it exists. The existing in-tree BlockTool
+        command is the priority over the same external plug-in command.
+        """
+        try:
+            if sys.version_info[0] == 2:
+                cmd_name = cmd_name.encode('ascii', 'replace')
+            mod = import_module('gnuradio.blocktool.cli.' + cmd_name)
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/base.py]: I can't see this working if we run from source, which we always said should be possible. `mod = import_module('.' + cmd_name, package=__package__)` would work, but I say get rid of this (see below)

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 26]
```diff
@@ -0,0 +1,301 @@
...
+
+
+@click.command('parseheader',
+               short_help='Generate the parsed output for the header file or directory in a specified format')
+@click.argument('file-path', nargs=1)
+@click.option('-yaml', '--yaml-confirm', is_flag=True,
+              help='If given, a YAML *file* alongside printed JSON response will be generated')
+@click.option('-c', '--blocktool-comments', is_flag=True,
+              help='blocktool helper comments will be added in the header file')
+@click.option('-f', '--file-confirm', is_flag=True,
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: How about -o/--output for file output instead of stdout

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 27]
```diff
@@ -0,0 +1,301 @@
...
+
+
+Dumper.add_representer(OrderedDict, dict_representer)
+Loader.add_constructor(_MAPPING_TAG, dict_constructor)
+
+
+@click.command('parseheader',
+               short_help='Generate the parsed output for the header file or directory in a specified format')
+@click.argument('file-path', nargs=1)
+@click.option('-yaml', '--yaml-confirm', is_flag=True,
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: --yaml to dump info as yaml instead of json

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 28]
```diff
@@ -0,0 +1,301 @@
...
+@click.option('-f', '--file-confirm', is_flag=True,
+              help='If given, file with default name in both JSON and YAML format will be generated')
+def cli(**kwargs):
+    """
+    \b
+    Get parsed output for a header file from GNU Radio module
+    """
+    kwargs['modtool'] = False
+    if os.path.isfile(kwargs['file_path']):
+        self = BlockHeaderParser(**kwargs)
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: self is not a good name here, how about parser?

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 29]
```diff
@@ -0,0 +1,301 @@
...
+        self.cli = True
+        self.json_confirm = True
+        run_blocktool(self)
+        click.secho('Header file: {}'.format(self.filename), fg='green')
+        json_generator(self, **kwargs)
+        if kwargs['yaml_confirm']:
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: I shoudl be either json or yaml

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 30]
```diff
@@ -0,0 +1,301 @@
...
+            self.yaml_confirm = True
+            yaml_generator(self)
+    elif os.path.isdir(kwargs['file_path']):
+        parse_directory(**kwargs)
+    else:
+        raise BlockToolException('Invalid file or directory path.')
+
+
+def json_generator(self, **kwargs):
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: An instance should call itself `self` not a function outside - rename to parser

Comment (skoslowski) [reply gr-utils/python/blocktool/cli/parseheader.py]: I also don't see a need for kwargs here

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 31]
```diff
@@ -0,0 +1,301 @@
...
+    """
+    Generate JSON file for the block header
+    """
+    header = self.filename.split('.')[0]
+    block = self.modname.split('-')[-1]
+    if kwargs['file_confirm']:
+        click.secho('Successfully generated {}_{}.json'.format(
+            block, header), fg='green')
+        json_file = os.path.join('.', block+'_'+header + '.json')
+        with open(json_file, 'w') as _file:
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: how about you pass in an open file handle for function to write too, this way it does not need to know the target (stdout or file on disk)

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 32]
```diff
@@ -0,0 +1,301 @@
...
+            json.dump(self.parsed_data, _file, indent=4)
+    else:
+        print(json.dumps(self.parsed_data, indent=4))
+        click.secho('Successfully parsed header file {}'.format(
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: we wanted the output to be json - so NO extra output...

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 33]
```diff
@@ -0,0 +1,301 @@
...
+
+
+def json_generator(self, **kwargs):
+    """
+    Generate JSON file for the block header
+    """
+    header = self.filename.split('.')[0]
+    block = self.modname.split('-')[-1]
+    if kwargs['file_confirm']:
+        click.secho('Successfully generated {}_{}.json'.format(
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: usually no output necessary on success for cli only warnings and errors - you can add a --verbose cli option if you want to keep those echos

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 34]
```diff
@@ -0,0 +1,301 @@
...
+        json_file = os.path.join('.', block+'_'+header + '.json')
+        with open(json_file, 'w') as _file:
+            json.dump(self.parsed_data, _file, indent=4)
+    else:
+        print(json.dumps(self.parsed_data, indent=4))
+        click.secho('Successfully parsed header file {}'.format(
+            self.filename), fg='green')
+
+
+def yaml_generator(self):
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: this seems to be the blocktool to grc-yaml code THAT BELONGS IN MODTOOL
the --yaml option in the cli should just dump the info in yaml format instead of json - as was discussed numerous times

File: gr-utils/python/blocktool/cli/parseheader.py [hunk 35]
```diff
@@ -0,0 +1,301 @@
...
+    cpp_templates = OrderedDict()
+    for tag, value in _cpp_templates:
+        cpp_templates[tag] = value
+    data['cpp_templates'] = cpp_templates
+
+    if self.parsed_data['docstring'] is not None:
+        data['documentation'] = self.parsed_data['docstring']
+    data['file_format'] = 1
+
+    with open(yml_file, 'w') as yml:
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/parseheader.py]: I will look at this function after you moved it to modtool.

File: gr-utils/python/blocktool/core/base.py [hunk 36]
```diff
@@ -0,0 +1,50 @@
...
+from __future__ import absolute_import
+from __future__ import unicode_literals
+
+
+class BlockToolException(Exception):
+    """ Standard exception for blocktool classes. """
+    pass
+
+
+class BlockTool(object):
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/base.py]: check out the abc module in python

File: gr-utils/python/modtool/cli/makeyaml.py [hunk 37]
```diff
@@ -24,14 +24,21 @@
...
 
+import os
 import click
 
 from ..core import get_block_candidates, ModToolMakeYAML
 from ..tools import SequenceCompleter
 from .base import common_params, block_name, run, cli_input
 
+from gnuradio.blocktool import BlockHeaderParser
+from gnuradio.blocktool.cli.base import BlockToolException, run_blocktool
```
Comment (skoslowski) [inline gr-utils/python/modtool/cli/makeyaml.py]: there must not be any usage of the blocktool cli here - use the API of the blocktool package here not its CLI. pretend it doesn't exist. if that means moving stuff into core then that is what has to happen

File: gr-utils/python/utils/gr_blocktool [hunk 38]
```diff
@@ -0,0 +1,32 @@
...
+
+
+from gnuradio.blocktool.cli.base import cli
+
+
+if __name__ == '__main__':
+    try:
+        cli()
+    except KeyboardInterrupt:
+        pass
```
Comment (skoslowski) [inline gr-utils/python/utils/gr_blocktool]: I say lets remove this in favor of blocktool.__main__

**Review by skoslowski — CHANGES_REQUESTED** [review]
This is the first bunch of comments - mostly focusing on the issues surround the core functionality - which is what I'll look at tomorrow in detail.

To be honest I am a bit surprised to find a lot of the issues we had discussed our calls still unsolved. Please address these. A part of me hopes I am looking at an outdated PR here...

File: gr-utils/python/blocktool/cli/parseheader.py (hunk 30 above)
Comment (aru31) [reply gr-utils/python/blocktool/cli/parseheader.py]: Required in line 101

Comment (aru31) [reply gr-utils/python/blocktool/cli/parseheader.py]: > I also don't see a need for kwargs here

Required in line 101

File: gr-utils/python/blocktool/cli/parseheader.py (hunk 27 above)
Comment (aru31) [reply gr-utils/python/blocktool/cli/parseheader.py]: okay, understood

File: gr-utils/python/blocktool/cli/parseheader.py (hunk 31 above)
Comment (aru31) [reply gr-utils/python/blocktool/cli/parseheader.py]: I think in case of parsing a directory this would create a problem.

Comment (skoslowski) [reply gr-utils/python/blocktool/cli/parseheader.py]: good point. nevermind then

File: gr-utils/python/blocktool/main.py [hunk 39]
```diff
@@ -0,0 +1,29 @@
...
+#
+""" A tool to parse the blocktool header files. """
+
+import sys
+from .cli.cli import cli
+
+
+def main():
+    """ Call the main function """
+    sys.exit(cli())
```
Comment (skoslowski) [inline gr-utils/python/blocktool/main.py]: I think this module can be merged into `__main__`, 
also (minor nitpick) `from .cli import cli should do here =)

File: gr-utils/python/blocktool/cli/cli.py [hunk 40]
```diff
@@ -0,0 +1,138 @@
...
+            except:
+                logging.basicConfig(level=logging.DEBUG,
+                                    filename=os.path.join('.', dir_name+'_log.out'))
+                logging.exception(
+                    'Log for Exception raised for the header: {}\n'.format(header))
+                click.secho('Parsing unsuccessful: {}'.format(
+                    header), fg='yellow')
+    else:
+        raise BlockToolException(
+            'Invalid directory! No header found to be parsed')
```
Comment (skoslowski) [inline gr-utils/python/blocktool/cli/cli.py]: since there is only one module left in the cli subpackage, lets make it a single module and move cli/cli.py to cli.py

File: gr-utils/python/blocktool/core/base.py [hunk 41]
```diff
@@ -0,0 +1,52 @@
...
+                 target_file=None, module=None, impldir=None, impl_file=None,
+                 yaml_confirm=False, json_confirm=False, cli=False, **kwargs):
+        """ __init__ """
+        pass
+
+    def _validate(self):
+        """ Validates the arguments """
+        pass
+
+    def run_blocktool(self):
```
Comment (skoslowski) [inline gr-utils/python/blocktool/core/base.py:51]: you went for an ABC, cool. But now this should be decorated with `@abc.abstractmethod` instead of raising a NotImplementedError
